core/src/test/java/org/infinispan/util/concurrent/BlockingManagerTest.java

```diff
@@ -0,0 +1,178 @@
+import java.util.concurrent.Executor;
+
+import org.infinispan.commons.test.BlockHoundHelper;
+import org.infinispan.test.AbstractInfinispanTest;
+import org.infinispan.test.Mocks;
+import org.mockito.Mockito;
+import org.reactivestreams.Publisher;
+import org.testng.annotations.Test;
+
+import io.reactivex.rxjava3.core.Flowable;
+import io.reactivex.rxjava3.processors.AsyncProcessor;
+import io.reactivex.rxjava3.processors.UnicastProcessor;
+import io.reactivex.rxjava3.subscribers.TestSubscriber;
+
+@Test(groups = "unit", testName = "util.concurrent.BlockingManagerTest")
+public class BlockingManagerTest extends AbstractInfinispanTest {
+   Executor nonBlockingExecutor;
+   Executor blockingExecutor;
+
+   public void initializeMocks() {
+      nonBlockingExecutor = Mockito.mock(Executor.class, Mockito.withSettings()
+            .defaultAnswer(Mocks.runWithExecutorAnswer(BlockHoundHelper.ensureNonBlockingExecutor())));
+      blockingExecutor = Mockito.mock(Executor.class, Mockito.withSettings()
+            .defaultAnswer(Mocks.runWithExecutorAnswer(BlockHoundHelper.allowBlockingExecutor())));
+   }
+
+   private BlockingManager createBlockingManager(boolean blockingInvocation) {
+      initializeMocks();
+
+      BlockingManagerImpl blockingManager = new BlockingManagerImpl() {
+         @Override
+         protected boolean isCurrentThreadBlocking() {
+            return blockingInvocation;
+         }
+      };
+
+      blockingManager.nonBlockingExecutor = nonBlockingExecutor;
+      blockingManager.blockingExecutor = blockingExecutor;
+      blockingManager.start();
+
+      return blockingManager;
+   }
+
+   public void testBlockingPublishToVoidStageInvokedBlockingThread() {
+      BlockingManager blockingManager = createBlockingManager(true);
+
+      CompletionStage<Void> stage = blockingManager.blockingPublisherToVoidStage(Flowable.fromArray(new Object[] { 1, 2, 3 })
+            .doOnNext(BlockHoundHelper::blockingConsume), null);
+      assertTrue(CompletionStages.isCompletedSuccessfully(stage));
+
+      // We should not have used any executor as we were a blocking thread already
+      Mockito.verifyZeroInteractions(nonBlockingExecutor, blockingExecutor);
+   }
+
+   public void testBlockingPublishToVoidStageInvokedNonBlockingThread() {
+      BlockingManager blockingManager = createBlockingManager(false);
+
+      CompletionStage<Void> stage = blockingManager.blockingPublisherToVoidStage(Flowable.just(1)
+            .doOnNext(BlockHoundHelper::blockingConsume), null);
+      assertTrue(CompletionStages.isCompletedSuccessfully(stage));
+
+      Mockito.verify(blockingExecutor).execute(Mockito.any());
+      Mockito.verifyZeroInteractions(nonBlockingExecutor);
+   }
+
+   public void testBlockingPublishToVoidStageInvokedNonBlockingThreadCompleteAfterSubscribe() {
+      BlockingManager blockingManager = createBlockingManager(false);
+
+      AsyncProcessor<Object> processor = AsyncProcessor.create();
```

Comment:
Is there any particular reason for using `AsyncProcessor` (compared to e.g. `UnicastProcess` or `PublishProcessor`) and for calling `processor.onNext(1)` before subscribing?

Reply:
> Is there any particular reason for using `AsyncProcessor` (compared to e.g. `UnicastProcess` or `PublishProcessor`)

Not in particular, but AsyncProcessor only allows a single value and the test only required one.

> and for calling `processor.onNext(1)` before subscribing?

Yes, this matters because if we do onNext before subscribing the subscribe "thread" will publish the value, where as if we called `onNext` afterwards the invoking thread would be the one publishing it.

Reply:
@wburns would it make sense to add more tests that call `processor.onNext(1)` afterwards, to test more scenarios?